include/SFML/Window/Event.hpp

```
@@ -34,196 +34,298 @@
-        SensorChanged,          //!< A sensor value changed (data in event.sensor)
-
-        Count //!< Keep last -- the total number of event types
+        Empty,                  //!< Default type
+        Closed,                 //!< The window requested to be closed
+        Resized,                //!< The window was resized
+        LostFocus,              //!< The window lost the focus
+        GainedFocus,            //!< The window gained the focus
+        TextEntered,            //!< A character was entered
+        KeyPressed,             //!< A key was pressed
+        KeyReleased,            //!< A key was released
+        MouseWheelScrolled,     //!< The mouse wheel was scrolled
+        MouseButtonPressed,     //!< A mouse button was pressed
+        MouseButtonReleased,    //!< A mouse button was released
+        MouseMoved,             //!< The mouse cursor moved
+        MouseEntered,           //!< The mouse cursor entered the area of the window
+        MouseLeft,              //!< The mouse cursor left the area of the window
+        JoystickButtonPressed,  //!< A joystick button was pressed
+        JoystickButtonReleased, //!< A joystick button was released
+        JoystickMoved,          //!< The joystick moved along an axis
+        JoystickConnected,      //!< A joystick was connected
+        JoystickDisconnected,   //!< A joystick was disconnected
+        TouchBegan,             //!< A touch event began
+        TouchMoved,             //!< A touch moved
+        TouchEnded,             //!< A touch event ended
+        SensorChanged,          //!< A sensor value changed
     };
 
+    Event() = default;
+
+    template <typename T>
+    explicit Event(const T& t) : m_data(t)
+    {
+    }
+
+    [[nodiscard]] Type getType() const
+    {
+        return static_cast<Type>(m_data.index());
+    }
+
+    template <typename T>
+    [[nodiscard]] bool is() const
```

For this (and the other similar members) I feel like there should be a safety to explicitly prevent anyone from using `sf::Event::Type::...` here. Haven't tried it yet, but this feels like it would compile and silently fail?

---

Just like the Event constructor this template ought to be constrained to just the types that exist in the underlying variant.